File: php/src/Simple/DeliveringOrder.php
<?php

namespace ImmutableDatamodel\Simple;

use DateTimeImmutable;
use ImmutableDatamodel\Shared\Address;
use ImmutableDatamodel\Shared\OrderConstraint;
use LogicException;

readonly class DeliveringOrder implements Order, OrderClosable
{
    /**
     * @param string $orderId
     * @param OrderConstraint[] $constraints
     * @param Address $deliveryAddress
     * @param DateTimeImmutable $deliveryDate
     */
    public function __construct(
        private string $orderId,
        private array $constraints,
        private Address $deliveryAddress,
        private DateTimeImmutable $deliveryDate
    ) {
        $this->validate($this->orderId, $this->constraints, $this->deliveryAddress, $this->deliveryDate);
    }

    private function validate(
        string $orderId,
        array $constraints,
        Address $deliveryAddress,
        DateTimeImmutable $deliveryDate
    ): void {
        if ($orderId == '') {
            throw new LogicException('This order id blank not allow');
        }

        if (in_array(OrderConstraint::SHIPPING_JAPAN_ONLY, $this->constraints)) {
            if ($this->deliveryAddress->country != 'JP') {
                throw new LogicException(OrderConstraint::SHIPPING_JAPAN_ONLY->getViolationMessage());
            }
        }

        if (in_array(OrderConstraint::DELIVERY_WEEKDAY, $this->constraints)) {
            if (in_array($this->deliveryDate->format('w'), ['0', '6'])) {
                throw new LogicException(OrderConstraint::DELIVERY_WEEKDAY->getViolationMessage());
            }
        }
    }

    public function orderId(): string
    {
        return $this->orderId;
    }

    public function constraints(): array
    {
        return $this->constraints;
    }

    public function close(): ClosedOrder
    {
        return new ClosedOrder($this->orderId(), $this->constraints());
    }
}

File: golang/src/simple/delivering_order.go
package simple

import (
	"errors"
	"example/shared"
	"time"
)

type DeliveringOrder struct {
	orderId         string
	constraints     []shared.OrderConstraint
	deliveryAddress shared.Address
	deliveryDate    time.Time
}

func NewDeliveringOrder(order Order, deliveryAddress shared.Address, deliveryDate time.Time) (DeliveringOrder, error) {
	if err := validate(order.OrderId(), order.Constraints(), deliveryAddress, deliveryDate); err != nil {
		return DeliveringOrder{}, err
	}
	return DeliveringOrder{
		orderId:         order.OrderId(),
		constraints:     order.Constraints(),
		deliveryAddress: deliveryAddress,
		deliveryDate:    deliveryDate,
	}, nil
}

func validate(
	orderId string,
	constraints []shared.OrderConstraint,
	deliveryAddress shared.Address,
	deliveryDate time.Time) error {
	if orderId == "" {
		return errors.New("order id blank not allow")
	}

	if deliveryAddress.Country == "" {
		return errors.New("order requires delivery address")
	}

	if deliveryDate.Year() == 1 {
		return errors.New("order requires delivery date")
	}

	contains := func(s []shared.OrderConstraint, need shared.OrderConstraint) bool {
		for _, v := range s {
			if v == need {
				return true
			}
		}
		return false
	}

	if contains(constraints, shared.SHIPPING_JAPAN_ONLY) {
		if deliveryAddress.Country != "JP" {
			return errors.New(shared.SHIPPING_JAPAN_ONLY.ViolationMessage())
		}
	}

	if contains(constraints, shared.DELIVERY_WEEKDAY) {
		if deliveryDate.Weekday() == time.Sunday ||
			deliveryDate.Weekday() == time.Saturday {
			return errors.New(shared.DELIVERY_WEEKDAY.ViolationMessage())
		}
	}
	return nil
}

func (do *DeliveringOrder) OrderId() string {
	return do.orderId
}

func (do *DeliveringOrder) Constraints() []shared.OrderConstraint {
	return do.constraints
}

func (do *DeliveringOrder) Close() ClosedOrder {
	return ClosedOrder{orderId: do.OrderId(), constraints: do.Constraints()}
}
